# Data Preparation and Customer Analytics
# Task 1

Conduct analysis on client's transaction dataset and identify customer purchasing behaviours to generate insights and provide commercial recommendations.

## Background Information for Task

We need to present a strategic recommendation to Julia that is supported by data which she can then use for the upcoming category review however to do so we need to analyse the data to understand the current purchasing trends and behaviours. The client is particularly interested in customer segments and their chip purchasing behaviour. Consider what metrics would help describe the customers’ purchasing behaviour.

## Work FLow

**1. Examine transaction data;** 

    - Look for inconsistencies
    - missing data across the dataset
    - outliers
    - correctly identify category items
    - numeric data across all tables.
    
    If anomalies are determined, make necessary changes and save it
    
**2. Examine customer data;**

    - Look for inconsistencies
    - missing data across the dataset
    - outliers
    - correctly identify category items
    - numeric data across all tables.

After completing the above tasks, merge the two datasets so it is ready for analysis.

**3. Data analysis and customer segments;**

    - define metrics
    - look at total sales
    - drivers of sales
    - where the highest sales are coming from
    - create charts and graphs
    - note interesting trends and insights
    
**4. Deep dive into customer segments;**

    - define recommendation from insights
    - determine which segments should be targeted
    - form overall conclusion based on analysis
    
**4. Save analysis**

    - save analysis and visuals

We will start this analysis by loading required libraries for analysis

In [1]:
# load libriaries
import numpy as np
import pandas as pd
import seaborn as sns 
import matplotlib.pyplot as plt
from scipy import stats
%matplotlib inline

In [2]:
# import data
transaction_df = pd.read_csv("../data/QVI_transaction_data.csv")
customer_df = pd.read_csv("../data/QVI_purchase_behaviour.csv")

## Exploratory Data Analysis

In [6]:
# view the first 10 rows of the data
transaction_df.head(10)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  43390          1            1000       1         5   
1  43599          1            1307     348        66   
2  43605          1            1343     383        61   
3  43329          2            2373     974        69   
4  43330          2            2426    1038       108   
5  43604          4            4074    2982        57   
6  43601          4            4149    3333        16   
7  43601          4            4196    3539        24   
8  43332          5            5026    4525        42   
9  43330          7            7150    6900        52   

                                  PROD_NAME  PROD_QTY  TOT_SALES  
0    Natural Chip        Compny SeaSalt175g         2        6.0  
1                  CCs Nacho Cheese    175g         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8  
5  Old El Paso Salsa   Dip Tomato Mild 300g         1        5.1  
6  Smiths Crinkle Chips Salt & Vinegar 330g         1        5.7  
7     Grain Waves         Sweet Chilli 210g         1        3.6  
8   Doritos Corn Chip Mexican Jalapeno 150g         1        3.9  
9     Grain Waves Sour    Cream&Chives 210G         2        7.2

In [7]:
# check the data types of each column
transaction_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 264836 entries, 0 to 264835
Data columns (total 8 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   DATE            264836 non-null  int64  
 1   STORE_NBR       264836 non-null  int64  
 2   LYLTY_CARD_NBR  264836 non-null  int64  
 3   TXN_ID          264836 non-null  int64  
 4   PROD_NBR        264836 non-null  int64  
 5   PROD_NAME       264836 non-null  object 
 6   PROD_QTY        264836 non-null  int64  
 7   TOT_SALES       264836 non-null  float64
dtypes: float64(1), int64(6), object(1)
memory usage: 16.2+ MB


We can observe that the `DATE` column has a data type of `int64` instead of a `datatime` format.

We will need to convert this from `int64` to `datetime`.

**Convert Data column to date time format**

In [8]:
transaction_df["DATE"]

0         43390
1         43599
2         43605
3         43329
4         43330
          ...  
264831    43533
264832    43325
264833    43410
264834    43461
264835    43365
Name: DATE, Length: 264836, dtype: int64

In [16]:
# import date and timedelta
from datetime import date, timedelta
start = date(year=1899, month=12, day=30)

transaction_df["DATE"] = transaction_df["DATE"].apply(lambda x: start + timedelta(x))

In [17]:
transaction_df["DATE"]

0         2018-10-17
1         2019-05-14
2         2019-05-20
3         2018-08-17
4         2018-08-18
             ...    
264831    2019-03-09
264832    2018-08-13
264833    2018-11-06
264834    2018-12-27
264835    2018-09-22
Name: DATE, Length: 264836, dtype: object

We have now converted the Dates to date time format.

We should check that we are looking at the right products by examining `PROD_NAME`

In [18]:
# get all the unique product names
transaction_df["PROD_NAME"].unique()

array(['Natural Chip        Compny SeaSalt175g',
       'CCs Nacho Cheese    175g',
       'Smiths Crinkle Cut  Chips Chicken 170g',
       'Smiths Chip Thinly  S/Cream&Onion 175g',
       'Kettle Tortilla ChpsHny&Jlpno Chili 150g',
       'Old El Paso Salsa   Dip Tomato Mild 300g',
       'Smiths Crinkle Chips Salt & Vinegar 330g',
       'Grain Waves         Sweet Chilli 210g',
       'Doritos Corn Chip Mexican Jalapeno 150g',
       'Grain Waves Sour    Cream&Chives 210G',
       'Kettle Sensations   Siracha Lime 150g',
       'Twisties Cheese     270g', 'WW Crinkle Cut      Chicken 175g',
       'Thins Chips Light&  Tangy 175g', 'CCs Original 175g',
       'Burger Rings 220g', 'NCC Sour Cream &    Garden Chives 175g',
       'Doritos Corn Chip Southern Chicken 150g',
       'Cheezels Cheese Box 125g', 'Smiths Crinkle      Original 330g',
       'Infzns Crn Crnchers Tangy Gcamole 110g',
       'Kettle Sea Salt     And Vinegar 175g',
       'Smiths Chip Thinly  Cut Original 175g', 'K

Looks like we are definitely looking at potato chips but how can we check that
these are all chips? We can do some basic text analysis by summarising the
individual words in the product name.

Examine the words in PROD_NAME to see if there are any incorrect entries such as products that are not chips

In [20]:
len(transaction_df["PROD_NAME"].unique())

114

We have 114 products in the data set

In [22]:
# check to see if all products ar chips
chip_product = []
for product in transaction_df["PROD_NAME"].unique():
    if "chip" in product.lower():
        chip_product.append(product)

print(len(chip_product))

33


Of the 114 unique products in the data set, only 33 are chip products

In [23]:
len(transaction_df["PROD_NAME"])

264836

In [27]:
transaction_df["PROD_NAME"].str.lower().str.contains("chip").sum()

74570

Of the 264836 in the rows 74570 are `chip` products.

In [35]:
# get the list of all the products names
productWords = list(transaction_df["PROD_NAME"].str.lower())
productWords

['natural chip        compny seasalt175g',
 'ccs nacho cheese    175g',
 'smiths crinkle cut  chips chicken 170g',
 'smiths chip thinly  s/cream&onion 175g',
 'kettle tortilla chpshny&jlpno chili 150g',
 'old el paso salsa   dip tomato mild 300g',
 'smiths crinkle chips salt & vinegar 330g',
 'grain waves         sweet chilli 210g',
 'doritos corn chip mexican jalapeno 150g',
 'grain waves sour    cream&chives 210g',
 'smiths crinkle chips salt & vinegar 330g',
 'kettle sensations   siracha lime 150g',
 'twisties cheese     270g',
 'ww crinkle cut      chicken 175g',
 'thins chips light&  tangy 175g',
 'ccs original 175g',
 'burger rings 220g',
 'ncc sour cream &    garden chives 175g',
 'doritos corn chip southern chicken 150g',
 'cheezels cheese box 125g',
 'smiths crinkle      original 330g',
 'ncc sour cream &    garden chives 175g',
 'infzns crn crnchers tangy gcamole 110g',
 'kettle sea salt     and vinegar 175g',
 'kettle tortilla chpshny&jlpno chili 150g',
 'old el paso salsa  

**Clean the `PROD_NAME` column**

As we are only interested in words that will tell us if the product is chips or
not, let's remove all words with digits and special characters such as '&' from our
set of product words

In [45]:
# Remove degits and special characters
productWords = list(transaction_df["PROD_NAME"].str.replace(r'([0-9]+[gG])','').str.replace(r'[^\w]',' ').str.lower())

C:\Users\Segun Lawal\AppData\Local\Temp\ipykernel_33768\100728002.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  productWords = list(transaction_df["PROD_NAME"].str.replace(r'([0-9]+[gG])','').str.replace(r'[^\w]',' ').str.lower())


In [46]:
productWords

['natural chip        compny seasalt',
 'ccs nacho cheese    ',
 'smiths crinkle cut  chips chicken ',
 'smiths chip thinly  s cream onion ',
 'kettle tortilla chpshny jlpno chili ',
 'old el paso salsa   dip tomato mild ',
 'smiths crinkle chips salt   vinegar ',
 'grain waves         sweet chilli ',
 'doritos corn chip mexican jalapeno ',
 'grain waves sour    cream chives ',
 'smiths crinkle chips salt   vinegar ',
 'kettle sensations   siracha lime ',
 'twisties cheese     ',
 'ww crinkle cut      chicken ',
 'thins chips light   tangy ',
 'ccs original ',
 'burger rings ',
 'ncc sour cream      garden chives ',
 'doritos corn chip southern chicken ',
 'cheezels cheese box ',
 'smiths crinkle      original ',
 'ncc sour cream      garden chives ',
 'infzns crn crnchers tangy gcamole ',
 'kettle sea salt     and vinegar ',
 'kettle tortilla chpshny jlpno chili ',
 'old el paso salsa   dip tomato mild ',
 'smiths chip thinly  cut original ',
 'kettle original ',
 'red rock deli thai 

**Most common words**

Let's look at the most common words by counting the number of times a word
appears and sorting them by this frequency in order of highest to lowest frequency

In [60]:
all_words = "".join( productWords).split()
all_words

['natural',
 'chip',
 'compny',
 'seasaltccs',
 'nacho',
 'cheese',
 'smiths',
 'crinkle',
 'cut',
 'chips',
 'chicken',
 'smiths',
 'chip',
 'thinly',
 's',
 'cream',
 'onion',
 'kettle',
 'tortilla',
 'chpshny',
 'jlpno',
 'chili',
 'old',
 'el',
 'paso',
 'salsa',
 'dip',
 'tomato',
 'mild',
 'smiths',
 'crinkle',
 'chips',
 'salt',
 'vinegar',
 'grain',
 'waves',
 'sweet',
 'chilli',
 'doritos',
 'corn',
 'chip',
 'mexican',
 'jalapeno',
 'grain',
 'waves',
 'sour',
 'cream',
 'chives',
 'smiths',
 'crinkle',
 'chips',
 'salt',
 'vinegar',
 'kettle',
 'sensations',
 'siracha',
 'lime',
 'twisties',
 'cheese',
 'ww',
 'crinkle',
 'cut',
 'chicken',
 'thins',
 'chips',
 'light',
 'tangy',
 'ccs',
 'original',
 'burger',
 'rings',
 'ncc',
 'sour',
 'cream',
 'garden',
 'chives',
 'doritos',
 'corn',
 'chip',
 'southern',
 'chicken',
 'cheezels',
 'cheese',
 'box',
 'smiths',
 'crinkle',
 'original',
 'ncc',
 'sour',
 'cream',
 'garden',
 'chives',
 'infzns',
 'crn',
 'crnchers',
 'tan

In [64]:
# get a count of the most frequent words in the product name
from collections import Counter

pd.Series(Counter(all_words)).sort_values(ascending=False)

chips           49770
kettle          38851
cheese          27890
smiths          26969
salt            24719
                ...  
chcknburger         8
chickenncc          8
saltgrnwves         8
htsunbites          8
saltsunbites        5
Length: 426, dtype: int64

We can see the most frequently occurring words and the least occurring words.

There are salsa products in the dataset but we are only interested in the chips
category, so let's remove these.

In [69]:
(~transaction_df["PROD_NAME"].str.lower().str.contains("salas")).sum()

264836

In [70]:
# get all entries that do not contain "salsa"
transaction_df["PROD_NAME"].str.lower().str.contains("salas")

0         False
1         False
2         False
3         False
4         False
          ...  
264831    False
264832    False
264833    False
264834    False
264835    False
Name: PROD_NAME, Length: 264836, dtype: bool

In [73]:
# Number of products that contains "salsa"
transaction_df["PROD_NAME"].str.lower().str.contains("salsa").sum()

18094

In [83]:
# number of products that do not contain "salsa"
(~transaction_df["PROD_NAME"].str.lower().str.contains("salsa")).sum()

246742

In [ ]:
transaction_df.shape

In [84]:
# products that do not contain "salsa"
transaction_df = transaction_df[~transaction_df["PROD_NAME"].str.lower().str.contains("salsa")]

In [85]:
transaction_df.shape

(246742, 8)